src/Material.cpp

```diff
@@ -663,38 +435,400 @@ void Material::setNumEnergyGroups(const int num_groups) {
+  
+  if (group <= 0 || group > _num_groups)
+    log_printf(ERROR, "Unable to get dif_coef for group %d for Material "
+               "%d which contains %d energy groups", group, _id, _num_groups);
+  
+  return _dif_coef[group-1];
+}
+
+
+/**
+ * @brief Get the Material's dif_hat for some energy group.
+ * @param group the energy group
+ * @param surface the index of the surface
+ * @return the dif_hat value
+ */
+FP_PRECISION MacroMaterial::getDifHatByGroup(int group, int surface) {
+  if (_dif_hat == NULL)
+    log_printf(ERROR, "Unable to return Material %d's dif_hat "
+               "since it has not yet been set", _id);
+  
+  if (group <= 0 || group > _num_groups)
+    log_printf(ERROR, "Unable to get dif_hat for group %d for Material "
+               "%d which contains %d energy groups", group, _id, _num_groups);
+               
+  return _dif_hat[surface*_num_groups + (group-1)];
+}
+
+
+/**
+ * @brief Get the Material's dif_tilde for some energy group.
+ * @param group the energy group
+ * @return the dif_tilde value
+ */
+FP_PRECISION MacroMaterial::getDifTildeByGroup(int group) {
+  if (_dif_tilde == NULL)
+    log_printf(ERROR, "Unable to return Material %d's dif_tilde "
+               "since it has not yet been set", _id);
+  
+  if (group <= 0 || group > _num_groups)
+    log_printf(ERROR, "Unable to get dif_tilde for group %d for Material "
+               "%d which contains %d energy groups", group, _id, _num_groups);
+  
+  return _dif_tilde[group-1];
+}
+
+
+/**
+ * @brief Get the Material's buckling for some energy group.
+ * @param group the energy group
+ * @return the buckling value
+ */
+FP_PRECISION MacroMaterial::getBucklingByGroup(int group) {
+  if (_buckling == NULL)
+    log_printf(ERROR, "Unable to return Material %d's buckling "
+               "since it has not yet been set", _id);
+  
+  if (group <= 0 || group > _num_groups)
+    log_printf(ERROR, "Unable to get buckling for group %d for Material "
+               "%d which contains %d energy groups", group, _id, _num_groups);
+  
+  return _buckling[group-1];
+}
+
+
+/**
+ * @brief Set the number of energy groups for this Material.
+ * @param num_groups the number of energy groups.
+ */
+void MacroMaterial::setNumEnergyGroups(const int num_groups) {
+  Material::setNumEnergyGroups(num_groups);
+  
+  /* Free old data arrays if they were allocated for a previous simulation */
+
+  /* If data is vector aligned */
+  if (_data_aligned) {
+    if (_sigma_s != NULL)
+      MM_FREE(_sigma_s);
+  }
+
+  /* Data is not vector aligned */
+  else {
+    if (_sigma_s != NULL)
+      delete [] _sigma_s;
+  }
+
+  if (_dif_coef != NULL)
+    delete [] _dif_coef;
+
+  if (_dif_hat != NULL)
+    delete [] _dif_hat;
+
+  if (_dif_tilde != NULL)
+    delete [] _dif_tilde;
+
+  if (_buckling != NULL)
+    delete [] _buckling;
+
+  /* Allocate memory for data arrays */
+  _sigma_s = new FP_PRECISION[_num_groups*_num_groups];
```

Comment:
The diffusion coefficients and buckling are only allocated if their getter or setter is called. Materials for the MOC solver don't necessarily need a diffusion coefficient or buckling. It they do have one, they are used to modify the absorption cross section. Furthermore the materials used for MOC don't need a `_dif_hat` or `_dif_tidle`; the materials used in the CMFD solver will need a `_dif_coef`, `_dif_hat`, and `_dif_tilde`.

One option to avoid any confusion would be to subclass `MacroMaterial` as `CMFDMaterial`, but that should probably be done in another PR.